File: test_sampling.cpp
/*
 * LibreSDR Sampling Test - C++ version
 * Verify IIO device can capture samples using libiio
 * 
 * Compile:
 *   g++ -o test_sampling test_sampling.cpp -liio -std=c++11
 * 
 * Run:
 *   ./test_sampling [uri] [duration_sec] [sample_rate_msps]
 *   Example: ./test_sampling ip:192.168.2.1 5 2.0
 */

#include <iio.h>
#include <iostream>
#include <string>
#include <vector>
#include <cstring>
#include <cstdlib>
#include <ctime>
#include <signal.h>
#include <unistd.h>
#include <limits>

static bool stop_capture = false;

void signal_handler(int sig) {
    std::cout << "\nInterrupted by user" << std::endl;
    stop_capture = true;
}

bool test_libresdr_sampling(const char* uri, int duration_sec, long long sample_rate) {
    std::cout << "LibreSDR Sampling Test (C++)" << std::endl;
    std::cout << std::string(60, '=') << std::endl;
    std::cout << "URI: " << uri << std::endl;
    std::cout << "Duration: " << duration_sec << " seconds" << std::endl;
    std::cout << "Sample rate: " << (sample_rate / 1e6) << " MSPS" << std::endl;
    std::cout << std::endl;
    
    // Create IIO context
    std::cout << "Connecting to device..." << std::endl;
    struct iio_context* ctx = iio_create_context_from_uri(uri);
    if (!ctx) {
        std::cerr << "✗ ERROR: Could not create IIO context for " << uri << std::endl;
        std::cerr << "  Check connection: ping 192.168.2.1" << std::endl;
        return false;
    }
    
    std::cout << "✓ Connected: " << iio_context_get_description(ctx) << std::endl;
    
    // Print context attributes
    unsigned int nb_attrs = iio_context_get_attrs_count(ctx);
    for (unsigned int i = 0; i < nb_attrs; i++) {
        const char* name;
        const char* value;
        if (iio_context_get_attr(ctx, i, &name, &value) == 0) {
            if (strcmp(name, "hw_model") == 0 || strcmp(name, "fw_version") == 0) {
                std::cout << "  " << name << ": " << value << std::endl;
            }
        }
    }
    std::cout << std::endl;
    
    // List devices
    std::cout << "Available IIO devices:" << std::endl;
    unsigned int nb_devices = iio_context_get_devices_count(ctx);
    for (unsigned int i = 0; i < nb_devices; i++) {
        struct iio_device* dev = iio_context_get_device(ctx, i);
        const char* name = iio_device_get_name(dev);
        unsigned int nb_channels = iio_device_get_channels_count(dev);
        std::cout << "  - " << (name ? name : iio_device_get_id(dev)) 
                  << ": " << nb_channels << " channels" << std::endl;
    }
    std::cout << std::endl;
    
    // Find RX device
    struct iio_device* rx_dev = iio_context_find_device(ctx, "cf-ad9361-lpc");
    if (!rx_dev) {
        std::cerr << "✗ ERROR: cf-ad9361-lpc device not found!" << std::endl;
        iio_context_destroy(ctx);
        return false;
    }
    
    std::cout << "Using RX device: " << (iio_device_get_name(rx_dev) ? 
              iio_device_get_name(rx_dev) : iio_device_get_id(rx_dev)) << std::endl;
    
    // Find PHY device for sample rate configuration
    struct iio_device* phy_dev = iio_context_find_device(ctx, "ad9361-phy");
    if (phy_dev) {
        struct iio_channel* phy_ch = iio_device_find_channel(phy_dev, "voltage0", false);
        if (phy_ch) {
            struct iio_channel_attr* attr = iio_channel_find_attr(phy_ch, "sampling_frequency");
            if (attr) {
                char buf[32];
                snprintf(buf, sizeof(buf), "%lld", sample_rate);
                if (iio_channel_attr_write(phy_ch, "sampling_frequency", buf) >= 0) {
                    long long actual_rate;
                    if (iio_channel_attr_read_longlong(phy_ch, "sampling_frequency", &actual_rate) == 0) {
                        std::cout << "✓ Sample rate set to: " << (actual_rate / 1e6) << " MSPS" << std::endl;
                    }
                }
            }
        }
    }
    
    // Enable RX channels (voltage0 and voltage1 for I/Q)
    std::vector<struct iio_channel*> rx_channels;
    for (const char* ch_name : {"voltage0", "voltage1"}) {
        struct iio_channel* ch = iio_device_find_channel(rx_dev, ch_name, false);
        if (ch) {
            iio_channel_enable(ch);
            rx_channels.push_back(ch);
            std::cout << "✓ Enabled channel: " << ch_name << std::endl;
        }
    }
    
    if (rx_channels.empty()) {
        std::cerr << "✗ ERROR: No RX channels could be enabled!" << std::endl;
        iio_context_destroy(ctx);
        return false;
    }
    std::cout << std::endl;
    
    // Create buffer
    size_t buffer_size = 16384;  // 16K samples
    std::cout << "Creating buffer (" << buffer_size << " samples)..." << std::endl;
    struct iio_buffer* buffer = iio_device_create_buffer(rx_dev, buffer_size, false);
    if (!buffer) {
        std::cerr << "✗ ERROR: Could not create buffer!" << std::endl;
        std::cerr << "  Hint: Increase buffer size on device:" << std::endl;
        std::cerr << "  ssh root@192.168.2.1 'echo 131072 > /sys/bus/iio/devices/iio:device3/buffer/length'" << std::endl;
        iio_context_destroy(ctx);
        return false;
    }
    std::cout << "✓ Buffer created: " << buffer_size << " samples" << std::endl;
    std::cout << std::endl;
    
    // Capture samples
    std::cout << "Capturing for " << duration_sec << " seconds..." << std::endl;
    std::cout << std::string(60, '-') << std::endl;
    
    signal(SIGINT, signal_handler);
    
    time_t start_time = time(NULL);
    unsigned long long total_samples = 0;
    unsigned int refill_count = 0;
    int16_t min_val = std::numeric_limits<int16_t>::max();
    int16_t max_val = std::numeric_limits<int16_t>::min();
    
    while (difftime(time(NULL), start_time) < duration_sec && !stop_capture) {
        ssize_t ret = iio_buffer_refill(buffer);
        if (ret < 0) {
            std::cerr << "\n✗ ERROR during buffer refill: " << ret << std::endl;
            break;
        }
        
        refill_count++;
        
        // Process each channel
        for (auto ch : rx_channels) {
            size_t sample_size = iio_channel_get_data_format(ch)->length / 8;
            size_t samples_in_buffer = iio_buffer_step(buffer) * 
                                       (iio_buffer_end(buffer) - iio_buffer_start(buffer)) / 
                                       iio_buffer_step(buffer);
            
            int16_t* data = (int16_t*)iio_buffer_first(buffer, ch);
            if (data) {
                for (size_t i = 0; i < samples_in_buffer; i++) {
                    int16_t val = data[i * iio_buffer_step(buffer) / sample_size];
                    if (val < min_val) min_val = val;
                    if (val > max_val) max_val = val;
                }
                total_samples += samples_in_buffer;
            }
        }
        
        // Progress
        double elapsed = difftime(time(NULL), start_time);
        double rate = (elapsed > 0) ? (total_samples / elapsed / 1e6) : 0;
        std::cout << "  Elapsed: " << elapsed << "s | Samples: " << total_samples 
                  << " | Rate: " << rate << " MSPS | Refills: " << refill_count
                  << " | Range: [" << min_val << ", " << max_val << "]        \r" << std::flush;
    }
    
    std::cout << std::endl;
    double elapsed = difftime(time(NULL), start_time);
    
    // Cleanup
    iio_buffer_destroy(buffer);
    iio_context_destroy(ctx);
    
    // Statistics
    std::cout << std::string(60, '-') << std::endl;
    std::cout << "\n📊 Capture Statistics:" << std::endl;
    std::cout << "  Duration: " << elapsed << " seconds" << std::endl;
    std::cout << "  Total samples: " << total_samples << std::endl;
    std::cout << "  Average rate: " << (total_samples / elapsed / 1e6) << " MSPS" << std::endl;
    std::cout << "  Buffer refills: " << refill_count << std::endl;
    std::cout << "  Sample range: [" << min_val << ", " << max_val << "]" << std::endl;
    
    // Validate
    std::cout << "\n✅ Validation:" << std::endl;
    bool success = true;
    
    if (total_samples == 0) {
        std::cout << "  ✗ FAIL: No samples captured!" << std::endl;
        success = false;
    } else {
        std::cout << "  ✓ PASS: Sample capture working" << std::endl;
    }
    
    if (min_val == max_val) {
        std::cout << "  ✗ FAIL: All samples same value (" << min_val << ")" << std::endl;
        success = false;
    } else {
        std::cout << "  ✓ PASS: Signal shows variation" << std::endl;
    }
    
    if (success) {
        std::cout << "\n" << std::string(60, '=') << std::endl;
        std::cout << "✅ SUCCESS: Device is capturing samples correctly!" << std::endl;
        std::cout << std::string(60, '=') << std::endl;
    }
    
    return success;
}

int main(int argc, char** argv) {
    const char* uri = "ip:192.168.2.1";
    int duration = 5;
    long long sample_rate = 2000000;  // 2 MSPS
    
    if (argc > 1) uri = argv[1];
    if (argc > 2) duration = atoi(argv[2]);
    if (argc > 3) sample_rate = (long long)(atof(argv[3]) * 1e6);
    
    bool success = test_libresdr_sampling(uri, duration, sample_rate);
    
    if (!success) {
        std::cout << "\n💡 Troubleshooting tips:" << std::endl;
        std::cout << "1. Check device connection: ping 192.168.2.1" << std::endl;
        std::cout << "2. Verify IIO device: iio_info -u ip:192.168.2.1" << std::endl;
        std::cout << "3. Increase buffer size on device" << std::endl;
        std::cout << "4. Check USB stability" << std::endl;
        return 1;
    }
    
    return 0;
}

File: test_sampling.py
#!/usr/bin/env python3
"""
LibreSDR Sampling Test - Verify IIO device can capture samples
Tests both RX channels and displays sample statistics
"""

import iio
import numpy as np
import time
import sys

def test_libresdr_sampling(uri="ip:192.168.2.1", duration_sec=5, sample_rate=2000000):
    """
    Test LibreSDR sampling capability
    
    Args:
        uri: IIO context URI (default: ip:192.168.2.1)
        duration_sec: How long to capture (seconds)
        sample_rate: Sample rate in Hz (default: 2 MSPS)
    """
    print(f"LibreSDR Sampling Test")
    print(f"=" * 60)
    print(f"URI: {uri}")
    print(f"Duration: {duration_sec} seconds")
    print(f"Sample rate: {sample_rate / 1e6:.1f} MSPS")
    print()
    
    try:
        # Create IIO context
        print("Connecting to device...")
        ctx = iio.Context(uri)
        print(f"✓ Connected: {ctx.description}")
        print(f"  Hardware: {ctx.attrs.get('hw_model', 'Unknown')}")
        print(f"  Firmware: {ctx.attrs.get('fw_version', 'Unknown')}")
        print()
        
        # Find devices
        print("Available IIO devices:")
        for dev in ctx.devices:
            print(f"  - {dev.name or dev.id}: {len(dev.channels)} channels")
        print()
        
        # Get RX device (cf-ad9361-lpc)
        rx_dev = ctx.find_device("cf-ad9361-lpc")
        if not rx_dev:
            print("✗ ERROR: cf-ad9361-lpc device not found!")
            return False
        
        print(f"Using RX device: {rx_dev.name or rx_dev.id}")
        print(f"Channels: {[ch.id for ch in rx_dev.channels]}")
        print()
        
        # Get PHY device for configuration
        phy_dev = ctx.find_device("ad9361-phy")
        if not phy_dev:
            print("✗ WARNING: ad9361-phy not found, cannot configure sample rate")
        else:
            # Set sample rate
            try:
                phy_dev.find_channel("voltage0", False).attrs["sampling_frequency"].value = str(sample_rate)
                actual_rate = int(phy_dev.find_channel("voltage0", False).attrs["sampling_frequency"].value)
                print(f"✓ Sample rate set to: {actual_rate / 1e6:.3f} MSPS")
            except Exception as e:
                print(f"✗ WARNING: Could not set sample rate: {e}")
                actual_rate = sample_rate
        
        # Configure RX channels (enable voltage0 and voltage1 for I/Q)
        rx_channels = []
        for ch_id in ["voltage0", "voltage1"]:
            try:
                ch = rx_dev.find_channel(ch_id, False)
                if ch:
                    ch.enabled = True
                    rx_channels.append(ch)
                    print(f"✓ Enabled channel: {ch_id}")
            except Exception as e:
                print(f"✗ WARNING: Could not enable {ch_id}: {e}")
        
        if len(rx_channels) == 0:
            print("✗ ERROR: No RX channels could be enabled!")
            return False
        
        print()
        
        # Create buffer
        buffer_size = 16384  # 16K samples
        print(f"Creating buffer ({buffer_size} samples)...")
        try:
            buffer = iio.Buffer(rx_dev, buffer_size)
            print(f"✓ Buffer created: {buffer_size} samples")
        except Exception as e:
            print(f"✗ ERROR: Could not create buffer: {e}")
            print(f"  Hint: Try increasing buffer size on device:")
            print(f"  ssh root@192.168.2.1 'echo 131072 > /sys/bus/iio/devices/iio:device3/buffer/length'")
            return False
        
        print()
        
        # Capture samples
        print(f"Capturing for {duration_sec} seconds...")
        print("-" * 60)
        
        start_time = time.time()
        total_samples = 0
        refill_count = 0
        min_val = float('inf')
        max_val = float('-inf')
        
        while (time.time() - start_time) < duration_sec:
            try:
                # Refill buffer (blocking call, waits for new data)
                buffer.refill()
                refill_count += 1
                
                # Read data from each channel
                for ch in rx_channels:
                    data = np.frombuffer(ch.read(), dtype=np.int16)
                    total_samples += len(data)
                    
                    if len(data) > 0:
                        min_val = min(min_val, data.min())
                        max_val = max(max_val, data.max())
                
                # Progress indicator
                elapsed = time.time() - start_time
                samples_per_sec = total_samples / elapsed if elapsed > 0 else 0
                print(f"  Elapsed: {elapsed:.1f}s | Samples: {total_samples:,} | "
                      f"Rate: {samples_per_sec / 1e6:.2f} MSPS | "
                      f"Refills: {refill_count} | Range: [{min_val}, {max_val}]", 
                      end='\r', flush=True)
                
            except KeyboardInterrupt:
                print("\n\nInterrupted by user")
                break
            except Exception as e:
                print(f"\n✗ ERROR during capture: {e}")
                return False
        
        print()  # New line after progress
        elapsed = time.time() - start_time
        
        # Statistics
        print("-" * 60)
        print("\n📊 Capture Statistics:")
        print(f"  Duration: {elapsed:.3f} seconds")
        print(f"  Total samples: {total_samples:,}")
        print(f"  Average rate: {total_samples / elapsed / 1e6:.3f} MSPS")
        print(f"  Buffer refills: {refill_count}")
        print(f"  Samples per refill: {total_samples / refill_count if refill_count > 0 else 0:.0f}")
        print(f"  Sample range: [{min_val}, {max_val}]")
        
        # Validate results
        print("\n✅ Validation:")
        if total_samples == 0:
            print("  ✗ FAIL: No samples captured!")
            return False
        elif total_samples < sample_rate * duration_sec * 0.5:
            print(f"  ⚠ WARNING: Captured fewer samples than expected")
            print(f"    Expected: ~{sample_rate * duration_sec / len(rx_channels):,.0f}")
            print(f"    Actual: {total_samples:,}")
        else:
            print(f"  ✓ PASS: Sample capture working correctly")
        
        if min_val == max_val:
            print(f"  ✗ FAIL: All samples have same value ({min_val}) - no signal variation!")
            return False
        elif abs(min_val) < 10 and abs(max_val) < 10:
            print(f"  ⚠ WARNING: Very low signal amplitude (range: {min_val} to {max_val})")
            print(f"    Check antenna connection and RF gain settings")
        else:
            print(f"  ✓ PASS: Signal shows variation (range: {min_val} to {max_val})")
        
        print("\n" + "=" * 60)
        print("✅ SUCCESS: Device is capturing samples correctly!")
        print("=" * 60)
        return True
        
    except Exception as e:
        print(f"\n✗ FATAL ERROR: {e}")
        import traceback
        traceback.print_exc()
        return False


if __name__ == "__main__":
    # Parse arguments
    uri = "ip:192.168.2.1"
    duration = 5
    sample_rate = 2000000  # 2 MSPS
    
    if len(sys.argv) > 1:
        uri = sys.argv[1]
    if len(sys.argv) > 2:
        duration = int(sys.argv[2])
    if len(sys.argv) > 3:
        sample_rate = int(float(sys.argv[3]) * 1e6)
    
    # Run test
    success = test_libresdr_sampling(uri, duration, sample_rate)
    
    if not success:
        print("\n💡 Troubleshooting tips:")
        print("1. Check device connection: ping 192.168.2.1")
        print("2. Verify IIO device: iio_info -u ip:192.168.2.1")
        print("3. Increase buffer size: ssh root@192.168.2.1 'echo 131072 > /sys/bus/iio/devices/iio:device3/buffer/length'")
        print("4. Check USB stability (disconnect/reconnect issues)")
        sys.exit(1)
    
    sys.exit(0)
